src/herder/HerderImpl.cpp

```diff
@@ -509,18 +509,46 @@ HerderImpl::processSCPQueue()
 void
 HerderImpl::processSCPQueueUpToIndex(uint64 slotIndex)
 {
-    while (true)
+    try
     {
-        SCPEnvelope env;
-        if (mPendingEnvelopes.pop(slotIndex, env))
-        {
-            getSCP().receiveEnvelope(env);
-        }
-        else
+        uint64 minSlotIndex = 0;
+        bool notTracking = !mHerderSCPDriver.trackingSCP();
+        while (true)
         {
-            return;
+            SCPEnvelope env;
+            if (mPendingEnvelopes.pop(minSlotIndex, slotIndex, env))
+            {
+                // ensures that regardless of what receiveEnvelope does, we do
+                // not process new messages the next time we loop if this would
+                // violate "tracking"'s invariant: in that we only process the
+                // current slot when tracking
+
+                // for lower slots, if messages for lower slots are queued for
```

Comment:
This reads as an incomplete sentence; is there something you wanted to say about the case of messages queued for lower slots?